File: app/src/main/java/fr/dmconcept/bob/models/BoardConfig.java
package fr.dmconcept.bob.models;

import java.util.ArrayList;

public class BoardConfig {

    private Long mId;
    private String mName;
    private ArrayList<ServoConfig> mServoConfigs;


    public BoardConfig(String name, ArrayList<ServoConfig> servoConfig){
        this(-1, name, servoConfig);
    }


    public BoardConfig(long id, String name, ArrayList<ServoConfig> servoConfigs){

        // Check that we got at least 1 servo config
        assert servoConfigs.size() > 0;

        this.mId           = id;
        this.mName         = name;
        this.mServoConfigs = servoConfigs;
    }

    public long getId(){
        return this.mId;
    }

    public String getName(){
        return this.mName;
    }

    public int getServoCount() {
        return this.mServoConfigs.size();
    }

    public ArrayList<ServoConfig> getServoConfigs(){
        return this.mServoConfigs;
    }

}

File: bob-server/src/fr/dmconcept/bob/common/models/BoardConfig.scala
package fr.dmconcept.bob.client.models

import fr.dmconcept.bob.client.models.json.BobJsonProtocol._
import java.util.UUID
import spray.json._

object BoardConfig {

  /**
   * Max quantity of servo per board
   */
  final val MAX_SERVOS = 7

  def apply(name: String, servoConfigs: Seq[ServoConfig]): BoardConfig = {
    BoardConfig(UUID.randomUUID.toString, name, servoConfigs)
  }

  def deserialize(json: JsValue) = json.convertTo[BoardConfig]

}

case class BoardConfig(
  id           : String,
  name         : String,
  servoConfigs : Seq[ServoConfig]
) {

  assert(!name.isEmpty        , "Empty name"         )
  assert(servoConfigs.nonEmpty, "Empty servo configs")

  {
    //Check that the servo configs don't contain duplicate pins
    val ports = servoConfigs.map(_.pin)
    assert(ports.size == ports.distinct.size, "Expected only distinct ports")
  }

}